zebra-chain/src/orchard/commitment.rs

```diff
@@ -0,0 +1,468 @@
+//! Note and value commitments.
+
+use std::{convert::TryFrom, fmt, io};
+
+use bitvec::prelude::*;
+use group::{prime::PrimeCurveAffine, GroupEncoding};
+use halo2::{
+    arithmetic::{CurveAffine, FieldExt},
+    pasta::pallas,
+};
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{
+    amount::{Amount, NonNegative},
+    serialization::{
+        serde_helpers, ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize,
+    },
+};
+
+use super::{
+    keys::{Diversifier, TransmissionKey},
+    note::Note,
+    sinsemilla::*,
+};
+
+/// Generates a random scalar from the scalar field 𝔽_{q_P}.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#pallasandvesta
+pub fn generate_trapdoor<T>(csprng: &mut T) -> pallas::Scalar
+where
+    T: RngCore + CryptoRng,
+{
+    let mut bytes = [0u8; 64];
+    csprng.fill_bytes(&mut bytes);
+    // pallas::Scalar::from_bytes_wide() reduces the input modulo q_P under the hood.
+    pallas::Scalar::from_bytes_wide(&bytes)
+}
+
+/// The randomness used in the Simsemilla Hash for note commitment.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct CommitmentRandomness(pallas::Scalar);
+
+/// Note commitments for the output notes.
+#[derive(Clone, Copy, Deserialize, PartialEq, Serialize)]
+pub struct NoteCommitment(#[serde(with = "serde_helpers::Affine")] pub pallas::Affine);
+
+impl fmt::Debug for NoteCommitment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // This will panic if the public key is the identity, which is bad news
+        // bears.
+        let (x, y) = self.0.get_xy().unwrap();
```

Comment:
Let's just return the string "identity"?

```suggestion
        let (x, y) = match self.0.get_xy() {
            Some((x, y)) => (x, y),
            None => return f.debug_tuple("NoteCommitment").field("identity").finish(),
        }
```